### PR TITLE
Implement TimeToSec function push down

### LINKED ISSUE #5116: Implement `TimeToSec` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `TimeToSec` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by hey-kong**
### What problem does this PR solve?

Issue Number: close #5116

Problem Summary:
Implement TimeToSec function push down.

### What is changed and how it works?
Related https://github.com/pingcap/tidb/pull/35739

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support function time_to_sec pushdown to tiflash.
```

### CODE REVIEW

File: dbms/src/Functions/tests/gtest_duration_pushdown.cpp [hunk 1]
```diff
@@ -166,5 +166,85 @@ try
...
+    auto & dur_data = dur_column->getData();
+    auto second_column = ColumnVector<Int64>::create();
+    auto & second_data = second_column->getData();
+    dur_data.resize(rowNum);
+    second_data.resize(rowNum);
+
+    std::random_device rd;
+    std::default_random_engine gen = std::default_random_engine(rd());
+    std::uniform_int_distribution<int> sign_dis(0, 1), hour_dis(0, 838), minute_dis(0, 59), second_dis(0, 59), microSecond_dis(0, 999999);
+    for (int i = 0; i < rowNum; i++)
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_duration_pushdown.cpp:233]:
```suggestion
    for (int i = 0; i < rowNum; ++i)
```

File: dbms/src/Functions/FunctionsDuration.cpp [hunk 2]
```diff
@@ -97,6 +97,57 @@ void FunctionDurationSplit<Impl>::executeImpl(Block & block, const ColumnNumbers
...
+    using FromFieldType = typename DataTypeMyDuration::FieldType;
+    const auto * col_from = checkAndGetColumn<ColumnVector<FromFieldType>>(block.getByPosition(arguments[0]).column.get());
+    if (col_from != nullptr)
+    {
+        const typename ColumnVector<FromFieldType>::Container & vec_from = col_from->getData();
+        const size_t size = vec_from.size();
+        auto col_to = ColumnVector<Int64>::create(size);
+        typename ColumnVector<Int64>::Container & vec_to = col_to->getData();
+
+        for (size_t i = 0; i < vec_from.size(); ++i)
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsDuration.cpp:135]: for (size_t i = 0; i < size; ++i)

**Review by ywqzzy — COMMENTED** [review]
Thanks for your contribution, nice job.

**Comment by zanmato1984** [comment]
Thanks for your contribution. We really appreciate that.

File: dbms/src/Common/MyDuration.cpp [hunk 3]
```diff
@@ -67,4 +67,10 @@ String MyDuration::toString() const
     auto frac_str = fmt::format("{:06}", microsecond);
     return fmt::format(fmt_str, sign > 0 ? "" : "-", hour, minute, second, frac_str);
 }
-} // namespace DB
\ No newline at end of file
+
+UInt64 calcSeconds(int hour, int minute, int second)
```
Comment (zanmato1984) [inline dbms/src/Common/MyDuration.cpp:71]: I think it's better to use `Int32` as the argument types. It should be the same as `MyDuration::hours()/minutes()/second()`'s return type.

File: dbms/src/Functions/FunctionsDuration.cpp [hunk 4]
```diff
@@ -97,6 +97,57 @@ void FunctionDurationSplit<Impl>::executeImpl(Block & block, const ColumnNumbers
             ErrorCodes::ILLEGAL_COLUMN);
 };
 
+template <typename Impl>
+DataTypePtr FunctionMyDurationToSec<Impl>::getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const
+{
+    if (!arguments[0].type->isMyTime())
+    {
+        throw Exception(
+            fmt::format("Illegal type {} of first argument of function {}", arguments[0].type->getName(), getName()),
```
Comment (zanmato1984) [inline dbms/src/Functions/FunctionsDuration.cpp:106]:
```suggestion
            fmt::format("Illegal type {} of the first argument of function {}", arguments[0].type->getName(), getName()),
```

File: dbms/src/Functions/FunctionsDuration.cpp [hunk 5]
```diff
@@ -97,6 +97,57 @@ void FunctionDurationSplit<Impl>::executeImpl(Block & block, const ColumnNumbers
...
+
+template <typename Impl>
+void FunctionMyDurationToSec<Impl>::executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const
+{
+    const auto * from_type = checkAndGetDataType<DataTypeMyDuration>(block.getByPosition(arguments[0]).type.get());
+    if (from_type == nullptr)
+    {
+        throw Exception(
+            fmt::format(
+                "Illegal column {} of first argument of function {}",
```
Comment (zanmato1984) [inline dbms/src/Functions/FunctionsDuration.cpp:120]:
```suggestion
                "Illegal column {} of the first argument of function {}",
```

File: dbms/src/Functions/FunctionsDuration.cpp [hunk 6]
```diff
@@ -97,6 +97,57 @@ void FunctionDurationSplit<Impl>::executeImpl(Block & block, const ColumnNumbers
...
+        {
+            MyDuration val(vec_from[i], from_type->getFsp());
+            vec_to[i] = Impl::apply(val);
+        }
+        block.getByPosition(result).column = std::move(col_to);
+    }
+    else
+        throw Exception(
+            fmt::format(
+                "Illegal column {} of first argument of function {}",
```
Comment (zanmato1984) [inline dbms/src/Functions/FunctionsDuration.cpp:145]:
```suggestion
                "Illegal column {} of the first argument of function {}",
```

File: dbms/src/Common/MyDuration.cpp [hunk 7]
```diff
@@ -67,4 +67,10 @@ String MyDuration::toString() const
...
+{
+    return hour * 3600 + minute * 60 + second;
```
Comment (zanmato1984) [inline dbms/src/Common/MyDuration.cpp:73]: There could be intermediate overflows during calculating `hour * 3600` or `minute * 60`. We should static cast them to `UInt64` first.

File: dbms/src/Common/MyDuration.cpp [hunk 8]
```diff
@@ -67,4 +67,10 @@ String MyDuration::toString() const
     auto frac_str = fmt::format("{:06}", microsecond);
     return fmt::format(fmt_str, sign > 0 ? "" : "-", hour, minute, second, frac_str);
 }
-} // namespace DB
\ No newline at end of file
+
+UInt64 calcSeconds(Int32 hour, Int32 minute, Int32 second)
```
Comment (zanmato1984) [inline dbms/src/Common/MyDuration.cpp:71]: This function is pretty `TimeToSec`-dedicated. Maybe we could move it to `FunctionsDuration.cpp` to minimize its scope. Or even discard it and write this single line of code inside `TiDBTimeToSecTransformerImpl::apply()` == manually inline it.

Comment (hey-kong) [reply dbms/src/Common/MyDuration.cpp]: ok

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Comment by zanmato1984** [comment]
@hey-kong Would you please link the corresponding tidb side pr in the description of this pr?

**Comment by hey-kong** [comment]
> @hey-kong Would you please link the corresponding tidb side pr in the description of this pr?

Done.

**Comment by zanmato1984** [comment]
> > @hey-kong Would you please link the corresponding tidb side pr in the description of this pr?
> 
> Done.

Thanks. I'm looking at it.

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Functions/FunctionsDuration.cpp [hunk 9]
```diff
@@ -97,6 +97,57 @@ void FunctionDurationSplit<Impl>::executeImpl(Block & block, const ColumnNumbers
             ErrorCodes::ILLEGAL_COLUMN);
 };
 
+template <typename Impl>
+DataTypePtr FunctionMyDurationToSec<Impl>::getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const
+{
+    if (!arguments[0].type->isMyTime())
```
Comment (bestwoody) [inline dbms/src/Functions/FunctionsDuration.cpp:103]: It seem type int, double and some strings can also work in MySQL.

```

mysql> SELECT TIME_TO_SEC(1234);
+-------------------+
| TIME_TO_SEC(1234) |
+-------------------+
|               754 |
+-------------------+
1 row in set (0.00 sec)

mysql> SELECT TIME_TO_SEC(12344);
+--------------------+
| TIME_TO_SEC(12344) |
+--------------------+
|               5024 |
+--------------------+
1 row in set (0.00 sec)

mysql> SELECT TIME_TO_SEC(12344.0);
+----------------------+
| TIME_TO_SEC(12344.0) |
+----------------------+
|                 5024 |
+----------------------+
1 row in set (0.00 sec)

mysql> SELECT TIME_TO_SEC(12344.1231231);
+----------------------------+
| TIME_TO_SEC(12344.1231231) |
+----------------------------+
|                       5024 |
+----------------------------+
1 row in set (0.00 sec)

mysql> SELECT TIME_TO_SEC("aasdf");
+----------------------+
| TIME_TO_SEC("aasdf") |
+----------------------+
|                 NULL |
+----------------------+
1 row in set, 1 warning (0.00 sec)

mysql> SELECT TIME_TO_SEC("123aasdf");
+-------------------------+
| TIME_TO_SEC("123aasdf") |
+-------------------------+
|                      83 |
+-------------------------+
1 row in set, 1 warning (0.00 sec)

```

Comment (zanmato1984) [reply dbms/src/Functions/FunctionsDuration.cpp:103]: I don't think we should be worrying about this. See https://github.com/pingcap/tidb/blob/e50b9430fc650ce4629ead05186272ffedf014c8/expression/builtin_time.go#L6047, there is only one `evalInt` method for `builtinTimeToSecSig`, meaning that `time_to_sec(int/double/string)` is handled by appended implicit casts, which will be propagated to TiFlash as well. So we can safely assume the operands of `time_to_sec` is a `duration`.

**Pull request merged**